Add basic support for Disting Ex dual mode commands

ryland opened the pull request:
Finally had a chance to connect disting to crow, so I went ahead and added these commands from the 1.9 firmware.

Not sure if the names for dual mode should be abbreviated-- they are getting a bit long.

tehn commented:
looks good--- don't have a disting so can't test but figure you have :)

@trentgill checking if you have opinions on name lengths

trentgill commented on lua/ii/disting.lua:
```diff
@@ -118,6 +118,56 @@ do return
+    , cmd = 0x5E
+    , docs = 'The 0-16384 value range scales to the actual param. value range'
+    , args = { { 'side_parameter', u8 }
+             , { 'value', u16 }
```
query if we can use `s16V` type here instead? in general we try and avoid scripters having to think in arbitrary scale ranges (eg 0-16384), instead opting for a more human friendly `volts` concept which is at least in base10.

the range [0..16384] is represented by [0.0 .. 10.0] with the `s16V` type. the underlying message that is sent is identical, so it's just a matter of the representation in a crow script.

i assume the values on disting itself are represented differently than as a raw 14bit value range -- if there is a numeric readout of values, i guess it would be best to mimic those values so one's script can speak in the same units as the module itself.

sorry if this is hard to grok - i've always had a hard time describing the custom type support in the i2c description lang.

discohead replied:
butting in here to note that the `s16v` values are kind of useless with disting i2c. To control disting via i2c you either use `set_parameter` and specify the literal parameter value or use `set_scale_parameter` / `set_controller` and send 0-16384. 0.0-10.0 is the arbitrary scale in this case.

trentgill replied:
ok i just looked a little at a video and maybe i get it now?

it seems like the 'parameter value' is not a unified range, but each different parameter has a different range (but the maximum range is 0 - 16384). if that's correct, then i guess there's nothing to do but force scripters to deal with the absolute range.

so correct me if i'm wrong: in *some* cases where the range is arbitrarily mapped to (0, 16384) it would be nice to use s16V, and even better in cases where the parameter is pitch, but some cases just use the numbers as indices or a smaller range, so you need to have precise integer values.

if that makes sense, then i agree s16V is not a good fit for an mutating param range like this. but please correct me if the above is not the issue.

discohead replied:
All disting parameters can be set in one of two ways, either you can specify the actual value that you want with no scaling or mapping whatsoever or you can send a 14-bit value and let the disting map it to the corresponding value in the set of valid values for that parameter. For example, on the Poly Wavetable synth algorithm in the Disting EX parameter 57 is a "Chorus mode" parameter and the possible values are 0, 1, or 2. So you can either use i2c command `0x46` to set parameter 57 to 0, 1 or 2... or you can use i2c command `0x47` to set parameter 57 via a 14-bit value that the disting will then map to mode 0, 1 or 2. In neither case is a range of 0.0..10.0 sensible. None of the i2c commands on the disting have any concept of "volts", so 10.0 will never be treated as a maximum unless it just so happens to be for that particular parameter. It's more like a 14-bit MIDI implementation. @scanner-darkly has posted the i2c specification here: https://github.com/scanner-darkly/teletype/wiki/DISTING-EX-I2C-SPECIFICATION

trentgill commented on lua/ii/disting.lua:
```diff
@@ -148,6 +198,46 @@ do return
     , retval = { 'state', u8 }
     , docs = '0-based loop index'
     }
+  , { name = 'dual_parameter'
```
wonder if we can abbreviate `parameter` to `param`? this obviously applies to all the above elements too.

similarly, could `algorithm` be `algo`?

trentgill commented on lua/ii/disting.lua:
```diff
@@ -148,6 +198,46 @@ do return
+
+if (data[0] >= 0x5A &&
+    data[0] <= 0x5C) {
+  data[1] = (data[1] & 0xF0) | (data[2] & 0xF);
+  data[1] = (data[1] & 0x0F) | ((side & 0xF) << 4);
```
i'm out of bit-manipulation practice, but couldn't this just be:
```c
data[1] = ((data[1] & 0xF) << 4) | (data[2] & 0xF);
```
then you could delete the `side` variable.

if there's something trickier going on, perhaps a short inline comment?

ryland replied:
I'm rusty on manipulation as well... hence a step based approach, but this is much more concise.

trentgill reviewed (commented):
all looks good to me overall. couple small inline thoughts & ideas, but it's in a good place already.

could you do a quick tab->spaces convert (there's a couple places where indentation gets off).

discohead commented on lua/ii/disting.lua:
```diff
@@ -118,6 +118,56 @@ do return
+        ,  { 'preset', u8 }
+      }
+    }
+  , { name = 'take_over_z'
```
in the disting manual the `0x65` command is described as `take-over/release Z`. it should be obvious to most disting users. The `z` knob on the disting can be overridden by MIDI or i2c.

ryland commented:
All the comments are good/helpful. I'll clean everything up when I have time in the next day or or two. Thanks @trentgill

ryland commented:
@trentgill I've cleaned this up a bit and am now handling all commands that use the packed side/param in pickle().

I removed all `set_` prefixes from command names in order to match convention, and removed `load_` where there was no corresponding `save_`. I think this helped enough that I didn't do any other abbreviations.

@discohead this works for me in some quick tests, but maybe you have time to try it out to confirm or offer any suggestions?
